File: src/sim.cpp

#include "sim.h"
#include "misc_v13.h"
#include "probability_v17.h"

using namespace std;

//------------------------------------------------
// Simulate from simple Wright-Fisher model
Rcpp::List sim_wrightfisher_cpp(Rcpp::List args, Rcpp::List args_functions, Rcpp::List args_progress) {
  
  // start timer
  chrono::high_resolution_clock::time_point t1 = chrono::high_resolution_clock::now();
  
  // get inputs from Rcpp format to base C++ format
  int N = rcpp_to_int(args("N"));
  int K = rcpp_to_int(args("K"));
  int L = rcpp_to_int(args("L"));
  vector<int> alleles = rcpp_to_vector_int(args("alleles"));
  double mu = rcpp_to_double(args("mu"));
  vector<vector<double>> mig_mat = rcpp_to_matrix_double(args("mig_mat"));
  vector<int> t_out = rcpp_to_vector_int(args("t_out"));
  int n_t_out = t_out.size();
  int initial_method = rcpp_to_int(args("initial_method"));
  vector<vector<double>> initial_params = rcpp_to_matrix_double(args("initial_params"));
  bool silent = rcpp_to_bool(args["silent"]);
  Rcpp::Function update_progress = args_functions["update_progress"];
  
  // objects for storing results
  vector<vector<vector<vector<int>>>> allele_counts_store(n_t_out);
  
  // initialise allele counts in each deme based on input method
  vector<vector<vector<int>>> allele_counts(K);
  if (initial_method == 1) {
    double theta = 2*N*mu;
    for (int k = 0; k < K; ++k) {
      allele_counts[k] = vector<vector<int>>(L);
      for (int l = 0; l < L; ++l) {
        vector<double> p = rdirichlet1(theta / (double)alleles[l], alleles[l]);
        allele_counts[k][l] = rmultinom1(N, p);
      }
    }
  } else if (initial_method == 2) {
    for (int k = 0; k < K; ++k) {
      allele_counts[k] = vector<vector<int>>(L);
      if (k == 0) {
        for (int l = 0; l < L; ++l) {
          vector<double> p = rdirichlet2(initial_params[l]);
          allele_counts[k][l] = rmultinom1(N, p);
        }
      } else {
        allele_counts[k] = allele_counts[0];
      }
    }
  }
  
  
  // create list of all pairwise demes that have positive migration rates
  vector<pair<pair<int, int>, double>> mig_list;
  for (int k1 = 0; k1 < (K - 1); ++k1) {
    for (int k2 = (k1 + 1); k2 < K; ++k2) {
      if (mig_mat[k1][k2] > 0) {
        pair<int, int> deme_pair = {k1, k2};
        mig_list.push_back(make_pair(deme_pair, mig_mat[k1][k2]));
      }
    }
  }
  
  // option to store initial values
  int t_out_next = 0;
  if (t_out[t_out_next] == 0) {
    allele_counts_store[t_out_next] = allele_counts;
    t_out_next++;
  }
  
  // loop through generations
  for (int t = 1; t < (max(t_out) + 1); ++t) {
    
    // report progress
    if (!silent) {
      update_progress(args_progress, "pb", t, max(t_out));
    }
    
    // apply migration
    for (int j = 0; j < mig_list.size(); ++j) {
      
      // draw number of migrants
      int n_mig = rbinom1(N, mig_list[j].second);
      
      // apply migration
      if (n_mig > 0) {
        pair<int, int> deme_pair = mig_list[j].first;
        int k1 = deme_pair.first;
        int k2 = deme_pair.second;
        
        for (int l = 0; l < L; ++l) {
          int tmp1 = n_mig;
          int tmp2 = n_mig;
          int tmp3 = N;
          int tmp4 = N;
          for (int i = 0; i < alleles[l]; ++i) {
           
            // subtract migrants from deme1
            int n_mig_i1 = rhyper1(allele_counts[k1][l][i], tmp3 - allele_counts[k1][l][i], tmp1);
            tmp1 -= n_mig_i1;
            tmp3 -= allele_counts[k1][l][i];
            allele_counts[k1][l][i] -= n_mig_i1;
            
            // subtract migrants from deme2
            int n_mig_i2 = rhyper1(allele_counts[k2][l][i], tmp4 - allele_counts[k2][l][i], tmp2);
            tmp2 -= n_mig_i2;
            tmp4 -= allele_counts[k2][l][i];
            allele_counts[k2][l][i] -= n_mig_i2;
            
            // add migrants
            allele_counts[k1][l][i] += n_mig_i2;
            allele_counts[k2][l][i] += n_mig_i1;
            
          }
        }
        
      }
    }
    
    // loop through demes and loci
    for (int k = 0; k < K; ++k) {
      for (int l = 0; l < L; ++l) {
        
        // apply drift by drawing from multinomial distribution with probabilities
        // given by previous generation frequencies
        int draws_remaining = N;
        int N_remaining = N;
        for (int i = 0; i < (alleles[l] - 1); ++i) {
          int count_new = rbinom1(draws_remaining, allele_counts[k][l][i] / (double)N_remaining);
          N_remaining -= allele_counts[k][l][i];
          draws_remaining -= count_new;
          allele_counts[k][l][i] = count_new;
          
          if (N_remaining == 0) {
            break;
          }
        }
        allele_counts[k][l][alleles[l] - 1] = draws_remaining;
        
        // apply mutation
        if (mu > 0) {
          int n_mut = rbinom1(N, mu);
          if (n_mut > 0) {
            
            // subtract mutants from existing counts
            int draws_remaining = n_mut;
            int N_remaining = N;
            for (int i = 0; i < (alleles[l] - 1); ++i) {
              int n_mut_i = rbinom1(draws_remaining, allele_counts[k][l][i] / (double)N_remaining);
              N_remaining -= allele_counts[k][l][i];
              draws_remaining -= n_mut_i;
              allele_counts[k][l][i] -= n_mut_i;
              
              if (N_remaining == 0) {
                break;
              }
            }
            allele_counts[k][l][alleles[l] - 1] -= draws_remaining;
            
            // add new mutants to allele counts
            draws_remaining = n_mut;
            for (int i = 0; i < (alleles[l] - 1); ++i) {
              int n_mut_i = rbinom1(draws_remaining, 1.0 / double(alleles[l] - i));
              draws_remaining -= n_mut_i;
              allele_counts[k][l][i] += n_mut_i;
            }
            allele_counts[k][l][alleles[l] - 1] += draws_remaining;
            
          }
        }
        
      }  // end loop through loci
    }  // end loop through demes
    
    // store result
    if (t == t_out[t_out_next]) {
      allele_counts_store[t_out_next] = allele_counts;
      if (t_out_next < (n_t_out - 1)) {
        t_out_next++;
      }
    }
    
  }  // end simulation loop
  
  // end timer
  chrono::high_resolution_clock::time_point t2 = chrono::high_resolution_clock::now();
  chrono::duration<double> time_span = chrono::duration_cast< chrono::duration<double> >(t2 - t1);
  print("simulation completed in", time_span.count(), "seconds\n");
  
  // return as Rcpp list
  Rcpp::List ret = Rcpp::List::create(Rcpp::Named("allele_counts") = allele_counts_store);
  return ret;
}

//------------------------------------------------
// apply box blur to matrix
Rcpp::NumericMatrix box_blur_cpp(Rcpp::NumericMatrix m, int d) {
  
  // initialise return matrix
  Rcpp::NumericMatrix ret(m.nrow(), m.ncol());
  int nr = ret.nrow();
  int nc = ret.ncol();
  
  // apply box blur to ret matrix
  for (int i = 0; i < nr; ++i) {
    for (int j = 0; j < nc; ++j) {
      
      // get limits of box
      int i_min = (i-d > 1) ? i-d : 1;
      int i_max = (i+d < nr) ? i+d : nr;
      int j_min = (j-d > 1) ? j-d : 1;
      int j_max = (j+d < nc) ? j+d : nc;
      
      // sum values within box
      double v = 0;
      for (int a = i_min; a < i_max; ++a) {
        for (int b = j_min; b < j_max; ++b) {
          v += m(a,b);
        }
      }
      
      // store mean value
      v /= double((i_max - i_min)*(j_max - j_min));
      ret(i,j) = v;
    }
  }
  
  return ret;
}

File: R/sim.R

#------------------------------------------------
#' @title Simulate allele frequencies from Wright-Fisher model
#'
#' @description Simulate Wright-Fisher evolution in a series of partially
#'   connected demes.
#'
#' @details Assumes a haploid population and independent loci (no linkage
#'   disequilibrium). Implements a finite-alleles mutation model with equal
#'   chance of mutating from any allele to any other. Migration is implemented
#'   by proposing random swaps of individuals between demes, thereby ensuring
#'   population sizes remain constant over time. For this reason, \code{N} must
#'   be the same for all demes.
#'   
#' @param N number of individuals per deme. Must be the same for all
#'   demes.
#' @param L number of loci (assumed independent).
#' @param alleles number of alleles. Can be a single number for all loci or a
#'   vector of length \code{L}.
#' @param mu mutation rate. Assumes finite-alleles model, with equal chance of
#'   mutating from any allele to any other.
#' @param mig_mat migration matrix specifying the per-generation probability of
#'   an individual migrating from any deme (in rows) to any other deme (in
#'   columns).
#' @param t_out vector of times at which results will be output.
#' @param initial_method,initial_params method of initialising allele
#'   frequencies, and parameters that are used in initialisation. There are two
#'   possible options:
#'   \enumerate{
#'   \item Each deme has allele frequencies drawn independently from a symmetric
#'   Dirichlet(theta/k) distribution, where \eqn{theta = 2*N*mu} and k is the
#'   number of alleles. This is the analytical equilibrium distribution under
#'   the model if there was no migration between demes.
#'     \item All demes have the same initial allele frequencies, which are drawn
#'     once from a Dirichlet(alpha_1, ..., alpha_k) distribution, where the
#'     alpha parameters are input as \code{initial_params}. This can be a vector
#'     if the same number of alleles is used over all loci, or a list of vectors
#'     over loci to accommodate varying numbers of alleles.
#'   }
#' @param silent if \code{TRUE} then suppress output to console.
#'
#' @importFrom utils txtProgressBar
#' @export

sim_wrightfisher <- function(N, L, alleles, mu, mig_mat, t_out,
                             initial_method = 1, initial_params = NULL, silent = FALSE) {
  
  # check inputs
  assert_single_pos_int(N, zero_allowed = FALSE)
  assert_single_pos_int(L, zero_allowed = FALSE)
  assert_vector_pos_int(alleles, zero_allowed = FALSE)
  assert_gr(alleles, 1)
  if (length(alleles) > 1) {
    assert_length(alleles, L)
  }
  assert_single_bounded(mu)
  assert_symmetric_matrix(mig_mat)
  assert_bounded(mig_mat)
  if (!all.equal(rowSums(mig_mat), rep(1, nrow(mig_mat)), check.attributes = FALSE)) {
    stop("every row of mig_mat must sum to 1")
  }
  assert_vector_pos_int(t_out, zero_allowed = TRUE)
  assert_in(initial_method, c(1, 2))
  if (initial_method == 2) {
    if (length(alleles) == 1) {
      assert_vector(initial_params)
      assert_length(initial_params, alleles)
    } else {
      assert_list(initial_params)
      assert_length(initial_params, L)
      for (i in 1:L) {
        assert_length(initial_params[[i]], alleles[i])
      }
    }
  }
  assert_single_logical(silent)
  
  # process some inputs
  if (length(alleles) == 1) {
    alleles <- rep(alleles, L)
  }
  if (initial_method == 2) {
    if (!is.list(initial_params)) {
      initial_params = replicate(L, initial_params, simplify = FALSE)
    }
  }
  
  # get number of demes from dimensions of migration matrix
  K <- ncol(mig_mat)
  
  # make argument list
  args <- list(N = N,
               K = K,
               L = L,
               alleles = alleles,
               mu = mu,
               mig_mat = matrix_to_rcpp(mig_mat),
               t_out = t_out,
               initial_method = initial_method,
               initial_params = initial_params,
               silent = silent)
  
  # create progress bars
  pb <- txtProgressBar(min = 0, max = max(c(1, t_out)), initial = NA, style = 3)
  args_progress <- list(pb = pb)
  
  # functions to pass to C++
  args_functions <- list(update_progress = update_progress)
  
  # run efficient C++ function
  output_raw <- sim_wrightfisher_cpp(args, args_functions, args_progress)
  
  # process output
  output_processed <- mapply(function(t_i) {
    mapply(function(k) {
      data.frame(time = t_out[t_i],
                 deme = k,
                 locus = rep(seq_len(L), times = alleles),
                 allele = unlist(lapply(alleles, seq_len)),
                 count = unlist(output_raw$allele_counts[[t_i]][[k]]))
    }, seq_len(K), SIMPLIFY = FALSE)
  }, seq_along(t_out), SIMPLIFY = FALSE) %>%
    dplyr::bind_rows()
  
  return(output_processed)
}


#------------------------------------------------
#' @title TODO
#'
#' @description TODO
#'   
#' @param project a genescaper project, as produced by the
#'   \code{genescaper_project()} function. The project must have a grid already
#'   defined (see \code{?create_hex_grid}) as this grid will be used in
#'   simulation.
#' @param loci number of loci (assumed independent).
#' @param alleles number of alleles, assumed to be the same over all loci.
#' @param nu,lambda,omega the values of the model parameters used in simulation.
#'   These parameters together specify the spatial autocorrelation function.
#' @param sigsq_shape,sigsq_rate,mu_mean,gamma parameters of the
#'   normal-inverse-gamma prior on the mean and variance of transformed allele
#'   frequencies.
#'
#' @importFrom stats rnorm
#'
#' @export

sim_GRF <- function(project, loci, alleles, nu, lambda, omega,
                    sigsq_shape, sigsq_rate, mu_mean, gamma) {
  
  # check inputs
  assert_class(project, "genescaper_project")
  assert_single_pos_int(loci, zero_allowed = FALSE)
  assert_single_pos_int(alleles, zero_allowed = FALSE)
  assert_single_bounded(nu)
  assert_single_pos(lambda, zero_allowed = FALSE)
  assert_single_bounded(omega, left = 1, right = 2)
  assert_single_pos(sigsq_shape, zero_allowed = FALSE)
  assert_single_pos(sigsq_rate, zero_allowed = FALSE)
  assert_single_numeric(mu_mean)
  assert_single_pos(gamma, zero_allowed = FALSE)
  
  # define correlation matrix
  centroids <- project$maps$grid$centroids
  n_site <- nrow(centroids)
  d <- get_GC_distance(centroids$longitude, centroids$latitude)
  K <- diag(1 - nu, n_site) + nu * exp(-(d / lambda)^omega)
  
  # draw means and variances
  n_sim <- loci * (alleles - 1)
  sigsq <- 1 / rgamma(n_sim, shape = sigsq_shape, rate = sigsq_rate)
  mu <- rnorm(n_sim, mean = mu_mean, sd = sqrt(sigsq / gamma))
  df_params <- data.frame(locus = rep(1:loci, each = alleles - 1),
                          allele = 1:(alleles - 1),
                          mu = mu,
                          sigsq = sigsq)
  
  # simulate z-values all in one go
  z <- mvtnorm::rmvnorm(n_sim, sigma = K) %>%
    sweep(1, mu, "+") %>%
    sweep(1, sqrt(sigsq), "*")
  
  # apply transformation to each locus
  z_list <- split(as.data.frame(z), f = df_params$locus)
  sim_wide <- mapply(function(i) {
    ret <- transform_z_to_p(t(z_list[[i]]))
    colnames(ret) <- sprintf("allele_%s", seq_len(ncol(ret)))
    ret %>%
      as.data.frame() %>%
      dplyr::mutate(site_ID = seq_len(nrow(ret)),
                    locus = i)
  }, seq_along(z_list), SIMPLIFY = FALSE) %>%
    dplyr::bind_rows()
  
  # get into long format
  sim_long <- sim_wide %>%
    tidyr::pivot_longer(-c(.data$site_ID, .data$locus), names_to = "allele", values_to = "freq") %>%
    dplyr::mutate(allele = as.numeric(as.factor(.data$allele)))
  
  return(list(data = sim_long,
              params = df_params))
}
